File: myTargetDemo/app/src/main/java/com/my/targetDemoApp/MainActivity.java
package com.my.targetDemoApp;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.my.targetDemoApp.fragments.PlusDialogFragment;
import com.my.targetDemoApp.models.AdvertisingType;
import com.my.targetDemoApp.utils.ActivityUtils;
import com.my.targetDemoApp.utils.MaterialColors;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class MainActivity extends AppCompatActivity implements PlusDialogFragment.SaveTypeListener
{
	private static final String KEY_STRING_SET = "saved_ads_set";
	private RecyclerView recyclerView;
	private SharedPreferences sharedPreferences;
	private MainAdapter adapter;
	private ArrayList<AdvertisingType> typeList;

	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);

		Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
		setSupportActionBar(toolbar);

		sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
		recyclerView = (RecyclerView) findViewById(R.id.main_grid);
		initGrid();
	}

	private void initGrid()
	{

		AdvertisingType defaultAd = new AdvertisingType(AdTypes.AD_TYPE_320X50, DefaultSlots.SLOT_STANDARD_BANNER);
		defaultAd.setName(getString(R.string.banners_320x50));
		defaultAd.setImageResource(R.drawable.img_banners);
		defaultAd.setDescription(getString(R.string.banners_320x50_desc));

		AdvertisingType interstitialAd = new AdvertisingType(AdTypes.AD_TYPE_FULLSCREEN, 0);
		interstitialAd.setName(getString(R.string.interstitial_ads));
		interstitialAd.setImageResource(R.drawable.img_interstitials);
		interstitialAd.setDescription(getString(R.string.interstitial_ads_desc));

		AdvertisingType nativeAd = new AdvertisingType(AdTypes.AD_TYPE_NATIVE, 0);
		nativeAd.setName(getString(R.string.native_ads));
		nativeAd.setImageResource(R.drawable.img_native);
		nativeAd.setDescription(getString(R.string.native_ads_desc));

		typeList = new ArrayList<>();
		typeList.add(defaultAd);
		typeList.add(interstitialAd);
		typeList.add(nativeAd);

		typeList.addAll(getSavedTypes());

		RecyclerView.ItemAnimator itemAnimator = new DefaultItemAnimator();
		itemAnimator.setAddDuration(300);
		itemAnimator.setRemoveDuration(300);
		recyclerView.setItemAnimator(itemAnimator);

		GridLayoutManager gridLayoutManager = new GridLayoutManager(getApplicationContext(),
				getResources().getInteger(R.integer.grid_coloumns));
		recyclerView.setLayoutManager(gridLayoutManager);


		adapter = new MainAdapter(typeList);
		recyclerView.setAdapter(adapter);
	}

	private ArrayList<AdvertisingType> getSavedTypes()
	{
		ArrayList<AdvertisingType> advertisingTypes = new ArrayList<>();

		Set<String> set = sharedPreferences.getStringSet(KEY_STRING_SET, new HashSet<String>());

		for (String s : set)
		{
			String[] sar = TextUtils.split(s, ":");
			AdvertisingType advertisingType = new AdvertisingType(sar[0], sar[1]);
			advertisingTypes.add(advertisingType);
		}
		return advertisingTypes;
	}

	@Override
	public void onSaveType(int adType, int slotId)
	{
		Set<String> set = sharedPreferences.getStringSet(KEY_STRING_SET, new HashSet<String>());
		set.add(adType + ":" + slotId);
		sharedPreferences.edit().remove(KEY_STRING_SET).apply();
		sharedPreferences.edit().putStringSet(KEY_STRING_SET, set).apply();
		AdvertisingType newType = new AdvertisingType(adType, slotId);
		typeList.add(newType);
		adapter.notifyItemInserted(typeList.size() - 1);
	}

	private void removeType(AdvertisingType type)
	{
		if (recyclerView.isAnimating()) return;
		int pos = typeList.indexOf(type);
		typeList.remove(pos);
		adapter.notifyItemRemoved(pos);
		Set<String> set = sharedPreferences.getStringSet(KEY_STRING_SET, new HashSet<String>());
		set.remove(type.getAdType() + ":" + type.getSlotId());
		sharedPreferences.edit().remove(KEY_STRING_SET).apply();
		sharedPreferences.edit().putStringSet(KEY_STRING_SET, set).apply();
	}

	private class MainAdapter extends
			RecyclerView.Adapter<MainAdapter.ViewHolder>
	{
		private final MaterialColors materialColors;
		private List<AdvertisingType> advertisingTypes;

		private MainAdapter(List<AdvertisingType> advertisingTypes)
		{
			this.advertisingTypes = advertisingTypes;
			materialColors = new MaterialColors();
		}

		@Override
		public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType)
		{
			View v = LayoutInflater.from(parent.getContext()).inflate(
					R.layout.adtype_card, parent, false);
			return new ViewHolder(v);
		}

		@Override
		public int getItemViewType(int position)
		{
			if (position == getItemCount() - 1)
				return 1;
			else return 0;
		}

		@Override
		public void onBindViewHolder(ViewHolder viewHolder, int position)
		{
			if (position == getItemCount() - 1)
			{
				viewHolder.nameLabel.setText(getString(R.string.custom_title));
				viewHolder.imageView.setImageResource(R.drawable.plus);
				viewHolder.descriptionLabel.setText(getString(R.string.custom_description));
				viewHolder.cardFrame.setBackgroundColor(Color.TRANSPARENT);
			} else
			{
				final AdvertisingType type = advertisingTypes.get(position);
				viewHolder.nameLabel.setText(type.getName());

				int res = type.getImageResource();
				if (res != 0)
				{
					viewHolder.imageView.setVisibility(View.VISIBLE);
					viewHolder.imageView.setImageResource(type.getImageResource());
					viewHolder.slotId.setVisibility(View.GONE);
					viewHolder.removeButton.setVisibility(View.GONE);
					viewHolder.cardFrame.setBackgroundColor(materialColors.get(position));
					viewHolder.descriptionLabel.setText(type.getDescription());
				} else
				{
					viewHolder.slotId.setVisibility(View.VISIBLE);
					viewHolder.imageView.setVisibility(View.GONE);
					viewHolder.nameLabel.setText(getResources()
							.getStringArray(R.array.advertisement_types)[type.getAdType()]);
					viewHolder.slotId.setText(String.valueOf(type.getSlotId()));
					viewHolder.removeButton.setVisibility(View.VISIBLE);
					viewHolder.cardFrame.setBackgroundColor(Color.LTGRAY);
					viewHolder.descriptionLabel.setText(getString(R.string.custom_ad));
					viewHolder.removeButton.setOnClickListener(new View.OnClickListener()
					{
						@Override
						public void onClick(View v)
						{
							removeType(type);
						}
					});
				}
			}
		}

		@Override
		public long getItemId(int position)
		{
			return position;
		}

		@Override
		public int getItemCount()
		{
			return advertisingTypes.size() + 1;
		}

		class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener
		{
			ImageView imageView;
			TextView nameLabel;
			TextView descriptionLabel;
			FrameLayout cardFrame;
			TextView slotId;
			ImageView removeButton;

			public ViewHolder(View itemView)
			{
				super(itemView);
				imageView = (ImageView) itemView.findViewById(R.id.iv_card_image);
				nameLabel = (TextView) itemView.findViewById(R.id.tv_card_title);
				descriptionLabel = (TextView) itemView.findViewById(R.id.tv_card_description);
				cardFrame = (FrameLayout) itemView.findViewById(R.id.card_frame);
				slotId = (TextView) itemView.findViewById(R.id.tv_slotid);
				removeButton = (ImageView) itemView.findViewById(R.id.iv_remove);

				itemView.setOnClickListener(this);
			}

			@Override
			public void onClick(View v)
			{
				int position = getLayoutPosition();
				if (position == getItemCount() - 1)
				{
					FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
					Fragment prev = getSupportFragmentManager().findFragmentByTag("dialog");
					if (prev != null)
					{
						ft.remove(prev);
					}
					ft.addToBackStack(null);

					PlusDialogFragment newFragment = new PlusDialogFragment();
					newFragment.setSaveTypeListener(MainActivity.this);
					newFragment.show(ft, "dialog");
				} else
				{
					if (advertisingTypes.size() > position && advertisingTypes.get(position) != null)
					{
						ActivityUtils.startNewActivity(MainActivity.this, advertisingTypes.get(position));
					}
				}
			}
		}
	}
}

File: myTargetDemo/app/src/main/java/com/my/targetDemoApp/MainActivity.kt
package com.my.targetDemoApp

import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import com.my.target.ads.MyTargetView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private lateinit var mainActivityAdapter: ItemsAdapter
    private lateinit var saver: Saver

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        saver = Saver(this)
        fab.setOnClickListener { addElement() }
        mainActivityAdapter = ItemsAdapter()
        main_recycler.adapter = mainActivityAdapter
        main_recycler.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        val callback = TouchCallback { mainActivityAdapter.deleteItem(it) }
        val helper = ItemTouchHelper(callback)
        helper.attachToRecyclerView(main_recycler)

        val types = ArrayList<ItemsAdapter.ListItem>()
        types.add(ItemsAdapter.ListItem({ goBanners() },
                                        getString(R.string.standard_banners),
                                        getString(R.string.standard_banners_desc)))
        types.add(ItemsAdapter.ListItem({ goInterstitials() },
                                        getString(R.string.interstitial_ads),
                                        getString(R.string.interstitial_ads_desc)))
        types.add(ItemsAdapter.ListItem({ goNative() },
                                        getString(R.string.native_ads),
                                        getString(R.string.native_ads_desc)))
        types.add(ItemsAdapter.ListItem({ goInstream() },
                                        getString(R.string.instream_ads),
                                        getString(R.string.instream_ads_desc)))
        callback.protectedTypesSize = types.size

        saver.restore()?.forEach { types.add(createListItem(it)) }

        mainActivityAdapter.setItems(types)
    }

    private fun goBanners(slot: Int? = null, adSize: Int? = null) {
        val intent = Intent(this, BannersActivity::class.java)
        if (slot != null && adSize != null) {
            intent.putExtra(BannersActivity.KEY_SLOT, slot)
            intent.putExtra(BannersActivity.KEY_SIZE, adSize)
        }
        startActivity(intent)
    }

    private fun goInterstitials() {
        startActivity(Intent(this, InterstitialsActivity::class.java))
    }

    private fun goNative(slot: Int? = null) {
        val intent = Intent(this, NativeActivity::class.java)
        if (slot != null) {
            intent.putExtra(NativeActivity.KEY_SLOT, slot)
        }
        startActivity(intent)
    }

    private fun goInstream(slot: Int? = null) {
        val intent = Intent(this, InstreamActivity::class.java)
        if (slot != null) {
            intent.putExtra(InstreamActivity.KEY_SLOT, slot)
        }
        startActivity(intent)
    }

    private fun goCustom(adType: CustomAdvertisingType) {
        val slot = adType.slotId ?: return
        when (adType) {
            CustomAdvertisingType.STANDARD_320X50  -> {
                goBanners(slot, MyTargetView.AdSize.BANNER_320x50)
            }
            CustomAdvertisingType.STANDARD_300X250 -> {
                goBanners(slot, MyTargetView.AdSize.BANNER_300x250)
            }
            CustomAdvertisingType.STANDARD_728X90  -> {
                goBanners(slot, MyTargetView.AdSize.BANNER_728x90)
            }
            CustomAdvertisingType.NATIVE           -> {
                goNative(slot)
            }
            CustomAdvertisingType.INSTREAM         -> {
                goInstream(slot)
            }
            CustomAdvertisingType.INTERSTITIAL     -> {
                val helper = InterstitialHelper(main_recycler)
                helper.init(slot, true)
            }

        }
    }

    private fun addElement() {
        val ft = supportFragmentManager.beginTransaction()
        val prev = supportFragmentManager.findFragmentByTag("dialog")
        if (prev != null) {
            ft.remove(prev)
        }
        ft.addToBackStack(null)

        val newFragment = PlusDialogFragment()
        newFragment.setSaveTypeListener { adType ->
            saver.save(adType)
            addCustomTypeToList(adType)
        }

        newFragment.show(ft, "dialog")
    }

    private fun addCustomTypeToList(adType: CustomAdvertisingType) {
        mainActivityAdapter.addItem(createListItem(adType))
    }

    private fun createListItem(adType: CustomAdvertisingType): ItemsAdapter.ListItem {
        return ItemsAdapter.ListItem({ goCustom(adType) },
                                     "Custom ${adType.toString().toLowerCase()}",
                                     "Slot ID ${adType.slotId}",
                                     {
                                         saver.remove(adType)
                                         mainActivityAdapter.deleteItem(it)
                                     })
    }

    class TouchCallback(private val swipeListener: (position: Int) -> Unit) : ItemTouchHelper.SimpleCallback(
            ItemTouchHelper.UP or ItemTouchHelper.DOWN,
            ItemTouchHelper.LEFT) {
        var protectedTypesSize: Int = 0

        override fun onMove(p0: RecyclerView,
                            p1: RecyclerView.ViewHolder,
                            p2: RecyclerView.ViewHolder): Boolean {
            return true
        }

        override fun onSwiped(p0: RecyclerView.ViewHolder, p1: Int) {
            swipeListener.invoke(p1)
        }

        override fun getSwipeDirs(recyclerView: RecyclerView,
                                  viewHolder: RecyclerView.ViewHolder): Int {
            return if (viewHolder.adapterPosition < protectedTypesSize) {
                0
            }
            else {
                super.getSwipeDirs(recyclerView, viewHolder)
            }
        }
    }
}
